cluster.go

```diff
@@ -68,47 +80,47 @@ func (c *ClusterClient) PoolStats() *PoolStats {
 		acc.TotalConns += m.TotalConns
 		acc.FreeConns += m.FreeConns
 	}
-	c.clientsMx.RUnlock()
 	return &acc
 }
 
+func (c *ClusterClient) closed() bool {
+	return atomic.LoadInt32(&c._closed) == 1
+}
+
 // Close closes the cluster client, releasing any open resources.
 //
 // It is rare to Close a ClusterClient, as the ClusterClient is meant
 // to be long-lived and shared between many goroutines.
 func (c *ClusterClient) Close() error {
-	defer c.clientsMx.Unlock()
-	c.clientsMx.Lock()
-
-	if c.closed {
+	if !atomic.CompareAndSwapInt32(&c._closed, 0, 1) {
 		return pool.ErrClosed
 	}
-	c.closed = true
+
+	c.clientsMx.Lock()
 	c.resetClients()
 	c.setSlots(nil)
+	c.clientsMx.Unlock()
```

Comment:
So slots is protected by its own mutex. If you don't want to use my `clientMap` suggestion above, I think you want to unlock the `clientMx` a line earlier (not that it matters much).